balancer.go

```diff
@@ -0,0 +1,329 @@
+// trigger hot-swap operation on running forwarder for pool and should
+// replace targets behind the load balancer without the restart
+func (lb *LoadBalancer) UpdatePool(pool ServicePool) error {
+	if len(pool.Identity()) == 0 {
+		return fmt.Errorf("pool missing identity")
+	}
+	lb.mutex.Lock()
+	defer lb.mutex.Unlock()
+	lb.poolMap[pool.Identity()] = pool
+	if fwd, exists := lb.forwarderMap[pool.Identity()]; exists {
+		fwd.UpdateServicePool(pool)
+	}
+	return nil
+}
+
+// Listen will try to launch balancer on all the required ports, strategy is all or nothing
+func (lb *LoadBalancer) Listen() error {
+
+	// Do this step to ensure that we will fail on misconfiguration if more than
+	// one service pool mapping presented to this load balance for mTLS
+	// TODO: TLS Allows manual verification for the handshake by that we can launch multiple pools on the same port
+	// TODO: for the first version we will limit 1 pool per port for simplicity
+	mapping, err := collectListenTargets(lb.poolMap)
+	if err != nil {
+		return err
+	}
+
+	type schedule struct {
+		port int
+		tls  *tls.Config
+		pool ServicePool
+	}
+	scheduleListeners := make([]schedule, len(mapping))
+	i := 0
+	// Build schedule list not to have thread failures at this stage
+	for port, identity := range mapping {
+
+		pki, err := tlsutil.FromPKI(identity.TLSData().GetCertificate(), identity.TLSData().GetPrivateKey())
+		if err != nil {
+			return fmt.Errorf("invalid service pool pki data")
+		}
+
+		config := &tls.Config{
+			Certificates: []tls.Certificate{pki.Certificate},
+			ClientAuth:   tls.RequestClientCert,
+			RootCAs:      pki.CertPool,
+			CipherSuites: []uint16{
+				tls.TLS_AES_128_GCM_SHA256,       // 2022 TLS v1.3 compliant
+				tls.TLS_CHACHA20_POLY1305_SHA256, // 2022 TLS v1.3 compliant
+				tls.TLS_AES_256_GCM_SHA384,       // 2022 TLS v1.3 compliant
+			},
+			MinVersion:       tls.VersionTLS13,
+			MaxVersion:       tls.VersionTLS13,
+			CurvePreferences: []tls.CurveID{tls.CurveP521, tls.CurveP384, tls.CurveP256},
+		}
+
+		scheduleListeners[i] = schedule{port, config, identity}
+		i++
+	}
+
+	// Schedule listeners one by one and fail if any of them fail
+	derCtx, derCancel := context.WithCancel(lb.runCtx)
+	errChan := make(chan error, len(scheduleListeners))
+	defer close(errChan)
+
+	wg := sync.WaitGroup{}
+	for _, params := range scheduleListeners {
+		wg.Add(1)
+		go func(ctx context.Context, cancelAll context.CancelFunc, errChan chan error, toSchedule schedule) {
+
+			// Don't forget to close all contexts
+			defer derCancel()
+			// Try to listen
+			listen, err := tls.Listen("tcp", fmt.Sprintf("localhost:%d", toSchedule.port), toSchedule.tls)
+			if err != nil {
+				errChan <- fmt.Errorf("failed to listen on port")
+				wg.Done()
+				return
+			}
+
+			// Spawn the coroutine to watch for the context break
+			go func(l net.Listener) {
+				// Here all required procedures were established for the listener thread
+				wg.Done()
```

Comment:
This `.Done()` should be deferred right ? Or am I misunderstanding here, as there's also not wg.Add prior to the goroutine ?

Reply:
Those are to defer all possible pools to try to start first (bind to ports) before decide to cancel them all if any of the ports unavailable (all or nothing).

Reply:
😓  It's definitely hard to grok this function. I figure that there must be a way to rearrange this into individual functions which are orchestrated by a top level function.